mapie/quantile_regression.py

```diff
@@ -0,0 +1,486 @@
+from __future__ import annotations
+from typing import Optional, Tuple, Union, cast, List
+
+import numpy as np
+from sklearn.base import RegressorMixin, clone
+from sklearn.linear_model import QuantileRegressor
+from sklearn.pipeline import Pipeline
+from sklearn.utils.validation import (
+    indexable,
+    check_is_fitted,
+    _num_samples,
+    _check_y,
+)
+from typing_extensions import TypedDict
+
+from ._typing import ArrayLike, NDArray
+from .utils import (
+    check_alpha_and_n_samples,
+    check_n_features_in,
+    check_null_weight,
+    fit_estimator,
+    check_lower_upper_bounds,
+)
+from ._compatibility import np_quantile
+from .regression import MapieRegressor
+
+
+class MapieQuantileRegressor(MapieRegressor):
+    """
+    This class implements the conformalized quantile regression strategy
+    as proposed by Romano et al. (2019) to make conformal predictions.
+    The only valid ``method`` is "quantile" and the only valid default
+    ``cv`` is "split".
+
+    Parameters
+    ----------
+    estimator : Optional[RegressorMixin]
+        Any regressor with scikit-learn API
+        (i.e. with fit and predict methods), by default ``None``.
+        If ``None``, estimator defaults to a ``QuantileRegressor`` instance.
+
+    method: str
+        Method to choose for prediction, in this case, the only valid method
+        is the "quantile" method.
+
+    cv: Optional[str]
+        The cross-validation strategy for computing residuals. The "split" cv
+        is the method used by default. It splits the dataset into a training
+        and calibration set.
+
+    alpha: float
+        Between 0 and 0.5, represents the risk level of the confidence
+        interval.
+        Lower ``alpha`` produce larger (more conservative) prediction
+        intervals.
+        ``alpha`` is the complement of the target coverage level.
+
+        By default 0.1.
+
+    Attributes
+    ----------
+    valid_methods: List[str]
+        List of all valid methods.
+
+    estimators_ : List[RegressorMixin]
+        - [0]: Estimator with quantile value of alpha/2
+        - [1]: Estimator with quantile value of 1 - alpha/2
+        - [2]: Estimator with quantile value of 0.5
+
+    conformity_scores_ : NDArray of shape (n_samples_train, 3)
+        Conformity scores between ``y_calib`` and ``y_pred``:
+            - [:, 0]: for y_calib coming from prediction estimator with
+            quantile of alpha/2
+            - [:, 0]: for y_calib coming from prediction estimator with
+            quantile of 1 - alpha/2
+            - [:, 2]: maximum of those first two scores
+
+    n_features_in_: int
+        Number of features passed to the fit method.
+
+    n_samples_calib: int
+        Number of samples in the calibration dataset.
+
+    References
+    ----------
+    Yaniv Romano, Evan Patterson and Emmanuel J. Candès.
+    "Conformalized Quantile Regression"
+    Advances in neural information processing systems 32 (2019).
+    """
```

Comment:
I would add a simple example here as in `MapieRegressor` (yes another one).

Reply:
done